### PR TITLE
Offline build support for Bzlmod

### PR DESCRIPTION
**Pull request by phoepsilonix**
Offline build support for Bzlmod
    
    Add bazel-central-registry to the sources.
    Add the --registry option to the bazel command.
    Add resolved.jq for update_mozc_deps
    
    bazel-7.1.1 -> bazel-7.3.1(github release)
    remove sdk-extensions
      - org.freedesktop.Sdk.Extension.llvm17
      - org.freedesktop.Sdk.Extension.bazel

1. In bazel-7.3.1, offline builds are possible using the --registry option and --distdir option together with bzlmod enabled.
1. In bazel-7.1.1, offline builds are possible with --noenable_bzlmod.
1. In bazel-7.1.1, the build itself seems to have no issues except for slight differences in the configuration of dependent modules.

Until bazel in flatpak is updated, we can build using --noenable_bzlmod, and once bazel is updated, we can transition to offline builds with bzlmod using the --registry option and --distdir option together.

### CODE REVIEW

**Comment by phoepsilonix** [comment]
Currently, it appears that network access is essential when --enable_bzlmod is enabled. It seems we have two options: either enable network access for flatpak-builder, or use the --noenable_bzlmod option, which would revert to a slightly older external module configuration.

**Comment by phoepsilonix** [comment]
It seems that offline builds are possible with bzlmod by preparing the Bazel Central Registry as a source and adding the --registry option.

**Comment by phoepsilonix** [comment]
It might not be possible with bazel-7.1.1. It seems that an update to the flatpak version of bazel is necessary.

**Comment by wengxt** [comment]
```
import http.server
import os.path
import urllib.request
import shutil
import socketserver

class BazelMirror(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        if not self.path or self.path.endswith("/"):
            self.send_error(http.HTTPStatus.NOT_FOUND)
            return None
        local_path = os.getcwd() + self.path
        if not os.path.isfile(local_path):
            directory = os.path.dirname(local_path)
            os.makedirs(directory, exist_ok=True)
            url = f'https:/{self.path}'
            print(f"download {url}")
            urllib.request.urlretrieve(url, local_path)
        with open(local_path, 'rb') as f:
            print(f"sending {local_path}")
            self.send_response(http.HTTPStatus.OK)
            self.send_header("Content-type", 'application/octet-stream')
            fs = os.fstat(f.fileno())
            self.send_header("Content-Length", str(fs[6]))
            self.send_header("Last-Modified",
                self.date_time_string(fs.st_mtime))
            self.end_headers()

            shutil.copyfileobj(f, self.wfile)
            if not self.wfile.closed:
                self.wfile.flush()
            self.rfile.close()
            print(f"done {local_path}")

class ThreadedHTTPServer(socketserver.ThreadingMixIn, http.server.HTTPServer):
    daemon_threads = True

server = ThreadedHTTPServer(
    ('127.0.0.1', 8000),
    BazelMirror
)

server.serve_forever()
```

```
rewrite (.*) http://127.0.0.1:8000/$1
```

Here is the other approach that I mentioned in the other comment that uses `--experimental_downloader_config=`

**Comment by phoepsilonix** [comment]
Good evening.
I see there's an option called --experimental_downloader_config.
While we don't know how Bazel's features will evolve in the future, if distdir is truly deprecated and there's no alternative functionality, this might be a necessary approach.

For now, since it seems it will take time for org.freedesktop.Sdk to update, I've tried a configuration where Bazel is installed from GitHub releases and built with gcc.

I think it should build without any issues.
Please check it out.

**Comment by wengxt** [comment]
> Good evening. I see there's an option called --experimental_downloader_config. While we don't know how Bazel's features will evolve in the future, if distdir is truly deprecated and there's no alternative functionality, this might be a necessary approach.
> 
> For now, since it seems it will take time for org.freedesktop.Sdk to update, I've tried a configuration where Bazel is installed from GitHub releases and built with gcc.
> 
> I think it should build without any issues. Please check it out.

1. both of our approach relies on a experimental flag, unfortunately. On that side, neither is optimal.
2. There is a issue with distdir: distdir is flat and if there are two files with same name is required, then it will have issue. In my test MODULE.bazel from different directory will be downloaded. This can be resolved with your --registry flag, so for now I guess we could still assume that there will be duplicate in the file name.

I think using --registry together with --distdir is a better option, since there is no experimental flag required during build time.

Ideally, we should use --registry with --repository_cache together. However the directory structure of repository_cache is also somehow a implementation details. I'm not fully convinced it won't change in the future. Looks like distdir is a better option for now.

However, I don't quiet like the way that you currently extract the downloading URL with experimental_repository_resolved_file .
1. it fully depends on bazel internal details and the schema is not fully documented.
2. experimental_repository_resolved_file generated file is "valid skylark", not "valid json". It works as json is.. by coincidence. skylark itself is closer to python. 
3. experimental_downloader_config is simpler comparing to experimental_repository_resolved_file.

In conclusion, I would prefer to:
1. use --registry & --distdir in flatpak manifest as you did in this change
2. In the manifest, https://github.com/bazelbuild/bazel-central-registry should use a fixed commit, instead of main to maintain the reproducibility. The commit can be updated within update_mozc_deps
3. generate required urls  & sha256 with experimental_downloader_config, instead of experimental_repository_resolved_file

**Comment by phoepsilonix** [comment]
I have made modifications for using the --experimental_downloader_config option.

1. In Ibus-Mozc, a dest-filename was necessary because icon names were duplicated, but in Fcitx5-Mozc, this consideration is no longer needed.
1. By using bcr, it seems we don't need to worry about duplicate names for remote patches either.
1. Also, we can expect that URLs mediated by the proxy will generally start with https.

Based on these factors, we have been able to significantly simplify the code.

**Comment by phoepsilonix** [comment]
I think most of the corrections are finished.

I'm also organizing things in a separate branch that I switched to from the local master branch.
Should I submit a new PR from that branch instead?
